trackintel/preprocessing/trips.py

```diff
@@ -0,0 +1,425 @@
+        belongs to (see notes below).
+
+    tours: GeoDataFrame (as trackintel tours)
+        The generated tours
+
+    Notes
+    -------
+    - Tours are defined as a collection of trips in a certain time frame that start and end at the same point
+    - Nested tours are possible and will be regarded as 2 (or more tours), where the smaller tour is also part of the
+      bigger tour. Therefore, one trip can belong to multiple tours! In the `tour_id` column of the trips table,
+      only the id of the smallest tour is used.
+    - It is possible to allow spatial gaps to occur on the tour, which might be useful to deal with missing data.
+      Example: The two trips home-work, supermarket-home would still be detected as a tour when max_nr_gaps >= 1,
+      although the work-supermarket trip is missing.
+      Warning: This only counts the number of gaps, but neither temporal or spatial distance of gaps, nor the number
+      of missing trips in a gap are bounded. Thus, this parameter should be set with caution, because trips that are
+      hours apart might still be connected to a tour if `max_nr_gaps > 0`.
+    """
+    # Two options: either the location IDs for staypoints on the trips are provided, or a maximum distance threshold
+    # between end and start of trips is used
+    if staypoints is not None:
+        assert (
+            "location_id" in staypoints.columns
+        ), "Staypoints with location ID is required, otherwise tours are generated without location using max_dist"
+        geom_col = None  # not used
+        crs_is_projected = False  # not used
+    else:
+        # if no location is given, we need the trips table to have a geometry column
+        assert isinstance(trips_inp, gpd.geodataframe.GeoDataFrame), "Trips table must be a GeoDataFrame"
+        geom_col = trips_inp.geometry.name
+        # get crs
+        crs_is_projected = ti.geogr.distances.check_gdf_crs(trips_inp.copy())
+
+    trips = trips_inp.copy()
+
+    kwargs = {
+        "max_dist": max_dist,
+        "max_nr_gaps": max_nr_gaps,
+        "max_time": max_time,
+        "staypoints": staypoints,
+        "geom_col": geom_col,
+        "crs_is_projected": crs_is_projected,
+    }
+    if print_progress:
+        tqdm.pandas(desc="User trip generation")
+        tours = (
+            trips.groupby(["user_id"], group_keys=False, as_index=False)
+            .progress_apply(_generate_tours_user, **kwargs)
+            .reset_index(drop=True)
+        )
+    else:
+        tours = (
+            trips.groupby(["user_id"], group_keys=False, as_index=False)
+            .apply(_generate_tours_user, **kwargs)
+            .reset_index(drop=True)
+        )
+
+    # No tours found
+    if len(tours) == 0:
+        return trips, tours
+
+    # index management
+    tours["id"] = np.arange(len(tours))
+    tours.set_index("id", inplace=True)
+
+    # assign tour id to trips
+    tour2trip_map = tours[["trips"]].to_dict()["trips"]
```

Comment:
Is it possible to do the id assignment using `explode`? like [here ](https://github.com/mie-lab/trackintel/blob/8372d269c9ab2489c3a4ef7a8bf4bb04625127bb/trackintel/preprocessing/triplegs.py#L244). I find that significantly increases readability. Again, we don't need to fix this in this PR :)

Reply:
Done - it really simplifies the code a lot!